### PR TITLE
Sweeping channels 

### PR DESCRIPTION
**Pull request by simzzz**
#73 and #113.

### CODE REVIEW

File: src/actions/identityActions.js [hunk 1]
```diff
@@ -240,11 +241,28 @@ export function identityWithdraw({ amountToWithdraw, withdrawTo }) {
 	return async function(dispatch, getState) {
 		try {
 			const account = selectAccount(getState())
+			const {
+				identityBalanceDai,
+				totalIdentityBalanceDai,
+			} = account.stats.formatteds
```
Comment (Ivshti) [inline src/actions/identityActions.js]: is that the right place to get this (cc @IvoPaunov @rori4 )?

File: src/actions/itemActions.js [hunk 2]
```diff
@@ -270,8 +271,24 @@ export function openCampaign({ campaign, account }) {
 				withBalance: [{ channel: campaign }],
 				account,
 			})
-
-			const { readyCampaign } = await openChannel({ campaign, account })
+			const { identityBalanceDai } = account.stats.formatted
```
Comment (Ivshti) [inline src/actions/itemActions.js]: again, is this the right place to get this number from?

File: src/actions/identityActions.js [hunk 3]
```diff
@@ -240,11 +241,28 @@ export function identityWithdraw({ amountToWithdraw, withdrawTo }) {
...
+			let sweepTxns
+			if (
```
Comment (Ivshti) [inline src/actions/identityActions.js]: do you need to check if it's less than `totalIdentityBalanceDai`? Shouldn't some other part of the code validate this (cc @IvoPaunov )?

File: src/services/smart-contracts/actions/core.js [hunk 4]
```diff
@@ -106,7 +111,134 @@ function getReadyCampaign(campaign, identity, Dai) {
 	return newCampaign
 }
 
-export async function openChannel({ campaign, account }) {
+async function getChannelsWithBalance({ identityAddr, requester, STATES }) {
+	const channels = await requester
+		.fetch({
+			route: '/campaigns',
```
Comment (Ivshti) [inline src/services/smart-contracts/actions/core.js]: you should use `?byEarner=` here

File: src/services/smart-contracts/actions/core.js [hunk 5]
```diff
@@ -106,7 +111,134 @@ function getReadyCampaign(campaign, identity, Dai) {
...
+
+	return Promise.all(
+		channels
+			.map(channel => {
+				const { lastApprovedBalances } = channel.status
+				if (lastApprovedBalances) {
+					const allLeafs = Object.keys(lastApprovedBalances).map(k =>
+						Channel.getBalanceLeaf(k, lastApprovedBalances[k])
+					)
+					const mTree = new MerkleTree(allLeafs)
```
Comment (Ivshti) [inline src/services/smart-contracts/actions/core.js]: I think you will need the `lastApprovedSigs` here; check the latest code in `routineAuthsLoop`

File: src/services/smart-contracts/actions/core.js [hunk 6]
```diff
@@ -106,7 +111,134 @@ function getReadyCampaign(campaign, identity, Dai) {
...
+				// Sorting by most balance so we can get top N needed using amountToSweep so we send as few transactions as possible
+				return new BN(c2.status.lastApprovedBalances[identityAddr]).gte(
+					new BN(c1.status.lastApprovedBalances[identityAddr])
+				)
+			})
+	)
+}
+
+async function getChannelsToSweepFrom({ amountToSweep, identityAddr }) {
+	const STATES = [
```
Comment (Ivshti) [inline src/services/smart-contracts/actions/core.js]: this is redundant; you can just get all states and not bother with the states (`?all`)

File: src/services/smart-contracts/actions/core.js [hunk 7]
```diff
@@ -106,7 +111,134 @@ function getReadyCampaign(campaign, identity, Dai) {
...
+					return channel.creator === identityAddr
+				}
+				return lastApprovedBalances && !!lastApprovedBalances[identityAddr]
+			})
+			.map(({ channel, lastApprovedBalances }) => ({
+				channel,
+				balance: lastApprovedBalances[identityAddr],
+			}))
+			.sort((c1, c2) => {
+				// Sorting by most balance so we can get top N needed using amountToSweep so we send as few transactions as possible
```
Comment (Ivshti) [inline src/services/smart-contracts/actions/core.js]: no, you have to sort by `outstanding`; and of course, calculate `outstanding` in the beginning

outstanding, for expired channels, is `channel.deposit - core.withdrawn(channel.id)` and for non-expired channels it's `lastApprovedBalances[identityAddr] - core.withdrawPerUser(channel.id, identityAddr)`

File: src/services/smart-contracts/actions/core.js [hunk 8]
```diff
@@ -106,7 +111,134 @@ function getReadyCampaign(campaign, identity, Dai) {
...
+	const identityAddr = account.identity.address
+	const channelsToSweep = await getChannelsToSweepFrom({
+		amountToSweep,
+		identityAddr,
+	})
+	const identityContract = new Contract(identityAddr, Identity.abi, provider)
+	const initialNonce = (await identityContract.nonce()).toNumber()
+
+	const txns = channelsToSweep.map((c, i) => {
+		const toWithdraw =
```
Comment (Ivshti) [inline src/services/smart-contracts/actions/core.js]: ok, so your `toWithdraw` is equivalent to the `outstanding` variable in `routineAuthsLoop`
also you need it when sorting but you don't need it here

`channelWithdrawExpired` just takes the entire channel 
`channelWithdraw` takes a bunch of stuff, one of them is the leaf from `lastApprovedBalances`

File: src/services/smart-contracts/actions/core.js (hunk 4 above)
Comment (IvoPaunov) [reply src/services/smart-contracts/actions/core.js]: that shouldn't be here at all, there is market service for that calls

File: src/services/smart-contracts/actions/core.js [hunk 9]
```diff
@@ -144,32 +136,21 @@ async function getChannelsWithBalance({ identityAddr, requester, STATES }) {
 			.map(({ channel, lastApprovedBalances }) => ({
 				channel,
 				balance: lastApprovedBalances[identityAddr],
+				outstanding:
+					channel.status.name === 'Expired'
+						? channel.deposit - Core.functions.withdrawn(channel.id)
```
Comment (Ivshti) [inline src/services/smart-contracts/actions/core.js]: aren't you missing an `await` here?

File: src/services/smart-contracts/actions/core.js [hunk 10]
```diff
@@ -144,32 +136,21 @@ async function getChannelsWithBalance({ identityAddr, requester, STATES }) {
...
+						: lastApprovedBalances[identityAddr] -
```
Comment (Ivshti) [inline src/services/smart-contracts/actions/core.js]: another missing `await`

File: src/actions/identityActions.js (hunk 1 above)
Comment (rori4) [reply src/actions/identityActions.js]: Isn't there a typo here? As far as I remember those came from `account.stats.formatted`

**Comment by IvoPaunov** [comment]
He will  change it to use .raw

**Comment by IvoPaunov** [comment]
some extra done:
- improved way to get identity nonce
- fixed extra components re-render
- added jazzicon (as metamask) for address visualization
- translation selector  `t` - just import `t` if you want `t`
- links to validator on  campaign details
- loading bar and label if there are no identities on metamask login
- add spinner selectors
- some bug fixes

**Pull request merged**